### ISSUE TITLE
Denormalization issues...

### ISSUE THREAD

**Issue by spacewalkingninja** [Y-debug-sys/Diffusion-TS#131]
Hey there!
I checked out your work and I am impressed. 
I have been training the model on top of my data to see how well it would work. 
I noticed an issue with the denormalization procedure though. 

Here is resulting issue:

<img width="2564" height="459" alt="Image" src="https://github.com/user-attachments/assets/6e74a2d4-9833-4a16-9e99-04c9fc0b0e01" />

I used both `neg_one_to_one = True `(I think it behaves better this way) and `neg_one_to_one = False`.
But still,.. I am not convinced everything is running good.
Here is my sampling workflow:

```
    args = parse_args()

    if args.seed is not None:
        seed_everything(args.seed)

    if args.gpu is not None:
        torch.cuda.set_device(args.gpu)
    
    config = load_yaml_config(args.config_file)
    config = merge_opts_to_config(config, args.opts)

    logger = Logger(args)
    logger.save_config(config)

    model = instantiate_from_config(config['model']).cuda()
    if args.sample == 1 and args.mode in ['infill', 'predict']:
        test_dataloader_info = build_dataloader_cond(config, args)
    dataloader_info = build_dataloader(config, args)
    trainer = Trainer(config=config, args=args, model=model, dataloader=dataloader_info, logger=logger)

    #if args.train:
    print("Train")
    try:
        trainer.train()
    except Exception:
        exit()
    print("Train completed")
    # Use this for conditional sampling (infill/predict)
    if args.sample == 1 and args.mode in ['infill', 'predict']:
        trainer.load(args.milestone)
        dataloader, dataset = test_dataloader_info['dataloader'], test_dataloader_info['dataset']
        coef = config['dataloader']['test_dataset']['coefficient']
        stepsize = config['dataloader']['test_dataset']['step_size']
        sampling_steps = config['dataloader']['test_dataset']['sampling_steps']

        # run restore (your existing call)
        samples, *_ = trainer.restore(dataloader, [dataset.window, dataset.var_num], coef, stepsize, sampling_steps)

        # optional: inverse transform to original range if needed
        #samples = inverse_transform_if_possible(samples, dataset)
        samples_denorm, den_info = denormalize(samples, dataset=dataset, save_dir=args.save_dir, verbose=True)
        print("Denormalize info:", den_info)
        samples = samples_denorm
        # if dataset contains ground truth / scaler for comparison, try to construct 'real'
        real = None
        try:
            if hasattr(dataset, 'samples') and not getattr(dataset, 'regular', True):
                # For test datasets where mask applied, dataset may contain 'samples' in normalized form
                if hasattr(dataset, 'scaler') and dataset.scaler is not None:
                    # apply scaler inverse if we can
                    real = dataset.samples
                    # attempt inverse map similar to samples
                    try:
                        real = unnormalize_to_zero_to_one(real)
                    except Exception:
                        if hasattr(dataset, 'scaler'):
                            b, sl, fn = real.shape
                            real = dataset.scaler.inverse_transform(real.reshape(-1, fn)).reshape(b, sl, fn)
        except Exception:
            real = None
        # also denormalize real if present
        if real is not None:
            try:
                real_den, info_real = denormalize(real, dataset=dataset, save_dir=args.save_dir, verbose=False)
                real = real_den
                print("Denormalized real using:", info_real)
            except Exception:
                pass
        # persist raw samples
        os.makedirs(args.save_dir, exist_ok=True)
        np.save(os.path.join(args.save_dir, f'ddpm_{args.mode}_{args.name}.npy'), samples)
        print(f"Saved samples to {os.path.join(args.save_dir, f'ddpm_{args.mode}_{args.name}.npy')}")

#AND THE DENORMALIZATION FUNCTION


def denormalize(samples, dataset=None, save_dir=None, verbose=True):
    """
    Try to convert samples (batch, seq_len, feat) back to original units.
    Priority:
      1) dataset.scaler (if present)
      2) scaler file found under save_dir (joblib .pkl)
      3) repo helper unnormalize_to_zero_to_one (if available) optionally followed by scaler
      4) return samples unchanged and warn
    """
    s = np.array(samples, copy=True)
    b, sl, f = s.shape
    info = {"used": None, "path": None}

    # Helper to apply scaler safely (handles [-1,1] -> [0,1] mapping if needed)
    def apply_scaler(sarr, scaler):
        flat = sarr.reshape(-1, f)
        # detect if values are in [-1,1] and scaler expects [0,1]
        try:
            fr = getattr(scaler, "feature_range", None)
        except Exception:
            fr = None
        vmin, vmax = flat.min(), flat.max()
        # If scaler expects 0..1 and data currently in [-1,1], map (x+1)/2
        if fr is not None and tuple(fr) == (0.0, 1.0) and vmin >= -1.1 and vmax <= 1.1:
            if verbose:
                print("Mapping data from [-1,1] to [0,1] before inverse_transform (scaler.feature_range==(0,1))")
            flat = (flat + 1.0) / 2.0
        # Some pipelines normalized to 0..1 already; in that case we don't remap
        try:
            inv = scaler.inverse_transform(flat)
            return inv.reshape(b, sl, f)
        except Exception as e:
            if verbose:
                print("scaler.inverse_transform failed:", e)
            raise

    # 1) dataset.scaler
    try:
        if dataset is not None and hasattr(dataset, "scaler") and dataset.scaler is not None:
            scaler = dataset.scaler
            den = apply_scaler(s, scaler)
            info["used"] = "dataset.scaler"
            return den, info
    except Exception as e:
        if verbose:
            print("dataset.scaler unusable:", e)

    # 2) saved scaler file under save_dir
    try:
        scaler, path = find_scaler_on_disk(save_dir)
        if scaler is not None:
            try:
                den = apply_scaler(s, scaler)
                info["used"] = "scaler_from_disk"
                info["path"] = path
                return den, info
            except Exception as e:
                if verbose:
                    print("scaler_from_disk apply failed:", e)
    except Exception as e:
        if verbose:
            print("error while searching for scaler on disk:", e)

    # 3) try repository helper unnormalize_to_zero_to_one if available
    try:
        if unnormalize_to_zero_to_one is not None:
            # if samples in [-1,1] this will map to [0,1] or otherwise behave as repo expects
            s_try = unnormalize_to_zero_to_one(s)
            # if we found a scaler on disk previously, try inverse-transform after this mapping
            scaler2, path2 = find_scaler_on_disk(save_dir)
            if scaler2 is not None:
                try:
                    den = apply_scaler(s_try, scaler2)
                    info["used"] = "unnormalize_to_zero_to_one + scaler_from_disk"
                    info["path"] = path2
                    return den, info
                except Exception:
                    pass
            # if no scaler, return the [0,1]-mapped result (maybe that's already original range)
            info["used"] = "unnormalize_to_zero_to_one_only"
            return s_try, info
    except Exception as e:
        if verbose:
            print("unnormalize_to_zero_to_one failed:", e)

    # 4) heuristic: if samples already look like real values, return them
    vmin, vmax = s.min(), s.max()
    if vmin < -1.1 or vmax > 1.1:
        info["used"] = "no_change_assumed_real"
        return s, info

    # 5) give up
    info["used"] = "failed_to_denorm"
    if verbose:
        print("Warning: failed to denormalize automatically. Provide dataset.scaler or a saved scaler file in", save_dir)
    return s, info

```

The config file is basically the Stocks config with dataset name changed. No changes there.
As you can see on the right side after the vertical line there are the "denormalized" results. 
They are totally OFF SCALE. 
There is some issue there, will appreciate some guidance :) !

**Comment by Y-debug-sys** [comment]
Thanks for sharing the details — I took a quick look at your setup and overall it seems fine. I’d suggest not performing the inverse normalization (or denormalization) to the original range immediately, and instead first inspect the normalized outputs to verify their value range and consistency.

**Comment by spacewalkingninja** [comment]
They look fine. I had started visualizing them like this at first. But for comparison with real data it is not exactly usable so I decided to apply the inverse normalization them . The values I get are what one can say is in the ball park but not there yet. For example real data of 191 vs denormalized output 184. I am not sure what is causing this..

**Comment by spacewalkingninja** [comment]
<img width="1499" height="632" alt="Image" src="https://github.com/user-attachments/assets/15f5105e-263c-45e0-aa40-8f0f593bfd50" />

Apparently the root of my issue was how the data is saved... Once I inspected the npy carefully I think I have got the hang of it!
I was plotting the first item from the npy when I need the last item (line 3475 from the screenshot) as it corresponds to the prediction. Correct me if I am wrong?

Ok!!! Just ran it exactly as I described and resutls are much more convincing!

<img width="2556" height="452" alt="Image" src="https://github.com/user-attachments/assets/eb8042b6-8f5d-4b57-a0b1-6a92b28460d0" />****
